File: SignalServiceKit/src/Util/OWS2FAManager.m
//
// Copyright 2018 Signal Messenger, LLC
// SPDX-License-Identifier: AGPL-3.0-only
//

#import "OWS2FAManager.h"
#import "AppReadiness.h"
#import "HTTPUtils.h"
#import "SSKEnvironment.h"
#import "TSAccountManager.h"
#import <SignalServiceKit/SignalServiceKit-Swift.h>

NS_ASSUME_NONNULL_BEGIN

NSString *const NSNotificationName_2FAStateDidChange = @"NSNotificationName_2FAStateDidChange";

NSString *const kOWS2FAManager_LastSuccessfulReminderDateKey = @"kOWS2FAManager_LastSuccessfulReminderDateKey";
NSString *const kOWS2FAManager_PinCode = @"kOWS2FAManager_PinCode";
NSString *const kOWS2FAManager_RepetitionInterval = @"kOWS2FAManager_RepetitionInterval";
NSString *const kOWS2FAManager_HasMigratedTruncatedPinKey = @"kOWS2FAManager_HasMigratedTruncatedPinKey";
NSString *const kOWS2FAManager_AreRemindersEnabled = @"kOWS2FAManager_AreRemindersEnabled";

const NSUInteger kHourSecs = 60 * 60;
const NSUInteger kDaySecs = kHourSecs * 24;

const NSUInteger kMin2FAPinLength = 4;
const NSUInteger kMin2FAv2PinLength = 4;
const NSUInteger kMax2FAv1PinLength = 20; // v2 doesn't have a max length
const NSUInteger kLegacyTruncated2FAv1PinLength = 16;

@interface OWS2FAManager ()

@property (nonatomic) OWS2FAMode mode;

@end

#pragma mark -

@implementation OWS2FAManager

#pragma mark -

+ (SDSKeyValueStore *)keyValueStore
{
    NSString *const kOWS2FAManager_Collection = @"kOWS2FAManager_Collection";
    return [[SDSKeyValueStore alloc] initWithCollection:kOWS2FAManager_Collection];
}

#pragma mark -

- (instancetype)init
{
    self = [super init];

    if (!self) {
        return self;
    }

    OWSSingletonAssert();

    AppReadinessRunNowOrWhenMainAppDidBecomeReadyAsync(^{
        if (self.mode == OWS2FAMode_V1) {
            OWSLogInfo(@"Migrating V1 reglock to V2 reglock");

            [self migrateToRegistrationLockV2]
                .done(^(id value) { OWSLogInfo(@"Successfully migrated to registration lock V2"); })
                .catch(^(NSError *error) {
                    OWSFailDebug(@"Failed to migrate V1 reglock to V2 reglock: %@", error.userErrorDescription);
                });
        }
    });

    return self;
}

- (nullable NSString *)pinCode
{
    __block NSString *_Nullable value;
    [self.databaseStorage
        readWithBlock:^(SDSAnyReadTransaction *transaction) { value = [self pinCodeWithTransaction:transaction]; }];
    return value;
}

- (nullable NSString *)pinCodeWithTransaction:(SDSAnyReadTransaction *)transaction
{
    return [OWS2FAManager.keyValueStore getString:kOWS2FAManager_PinCode transaction:transaction];
}

- (void)setPinCode:(nullable NSString *)pin transaction:(SDSAnyWriteTransaction *)transaction
{
    if (pin.length == 0) {
        [OWS2FAManager.keyValueStore removeValueForKey:kOWS2FAManager_PinCode transaction:transaction];
        return;
    }

    if (self.hasBackedUpMasterKey) {
        pin = [KeyBackupServiceObjcBridge normalizePin:pin];
    } else {
        // Convert the pin to arabic numerals, we never want to
        // operate with pins in other numbering systems.
        pin = pin.ensureArabicNumerals;
    }

    [OWS2FAManager.keyValueStore setString:pin key:kOWS2FAManager_PinCode transaction:transaction];
}

- (OWS2FAMode)mode
{
    // Identify what version of 2FA we're using
    if (self.hasBackedUpMasterKey) {
        return OWS2FAMode_V2;
    } else if (self.pinCode != nil) {
        return OWS2FAMode_V1;
    } else {
        return OWS2FAMode_Disabled;
    }
}

- (BOOL)is2FAEnabled
{
    return self.mode != OWS2FAMode_Disabled;
}

- (void)markDisabledWithTransaction:(SDSAnyWriteTransaction *)transaction
{
    [OWS2FAManager.keyValueStore removeValueForKey:kOWS2FAManager_PinCode transaction:transaction];
    [OWS2FAManager.keyValueStore removeValueForKey:OWS2FAManager.isRegistrationLockV2EnabledKey transaction:transaction];

    [transaction addSyncCompletion:^{
        [[NSNotificationCenter defaultCenter] postNotificationNameAsync:NSNotificationName_2FAStateDidChange
                                                                 object:nil
                                                               userInfo:nil];

        [self.tsAccountManager updateAccountAttributes].catch(^(NSError *error) {
            OWSLogError(@"Error: %@", error);
        });
    }];
}

- (void)markEnabledWithPin:(NSString *)pin transaction:(SDSAnyWriteTransaction *)transaction
{
    [self setPinCode:pin transaction:transaction];

    // Since we just created this pin, we know it doesn't need migration. Mark it as such.
    [self markLegacyPinAsMigratedWithTransaction:transaction];

    // Reset the reminder repetition interval for the new pin.
    [self setDefaultRepetitionIntervalWithTransaction:transaction];

    // Schedule next reminder relative to now
    [self setLastCompletedReminderDate:[NSDate new] transaction:transaction];

    [transaction addSyncCompletion:^{
        [[NSNotificationCenter defaultCenter] postNotificationNameAsync:NSNotificationName_2FAStateDidChange
                                                                 object:nil
                                                               userInfo:nil];

        [self.tsAccountManager updateAccountAttributes].catch(^(NSError *error) {
            OWSLogError(@"Error: %@", error);
        });
    }];
}

- (void)requestEnable2FAWithPin:(NSString *)pin
                           mode:(OWS2FAMode)mode
                rotateMasterKey:(BOOL)rotateMasterKey
                        success:(nullable OWS2FASuccess)success
                        failure:(nullable OWS2FAFailure)failure
{
    OWSAssertDebug(pin.length > 0);
    OWSAssertDebug(success);
    OWSAssertDebug(failure);

    switch (mode) {
        case OWS2FAMode_V2: {
            // Enabling V2 2FA doesn't inherently enable registration lock,
            // it's managed by a separate setting.
            [self generateAndBackupKeysWithPin:pin rotateMasterKey:rotateMasterKey]
                .done(^(id value) {
                    OWSAssertIsOnMainThread();

                    DatabaseStorageWrite(self.databaseStorage, ^(SDSAnyWriteTransaction *transaction) {
                        [self markEnabledWithPin:pin transaction:transaction];
                    });

                    if (success) {
                        success();
                    }
                })
                .catch(^(NSError *error) {
                    OWSAssertIsOnMainThread();

                    if (failure) {
                        failure(error);
                    }
                });
            break;
        }
        case OWS2FAMode_V1:
            [self enable2FAV1WithPin:pin success:success failure:failure];
            break;
        case OWS2FAMode_Disabled:
            OWSFailDebug(@"Unexpectedly attempting to enable 2fa for disabled mode");
            break;
    }
}

- (void)disable2FAWithSuccess:(nullable OWS2FASuccess)success failure:(nullable OWS2FAFailure)failure
{
    switch (self.mode) {
        case OWS2FAMode_V2:
        {
            [self deleteKeys]
                .then(^(id value) { return [self disableRegistrationLockV2]; })
                .ensure(^{
                    OWSAssertIsOnMainThread();

                    DatabaseStorageWrite(self.databaseStorage, ^(SDSAnyWriteTransaction *transaction) {
                        [self markDisabledWithTransaction:transaction];
                    });
                })
                .done(^(id value) {
                    OWSAssertIsOnMainThread();

                    if (success) {
                        success();
                    }
                })
                .catch(^(NSError *error) {
                    OWSAssertIsOnMainThread();

                    if (failure) {
                        failure(error);
                    }
                });
            break;
        }
        case OWS2FAMode_V1:
            [self disable2FAV1WithSuccess:success failure:failure];
            break;
        case OWS2FAMode_Disabled:
            OWSFailDebug(@"Unexpectedly attempting to disable 2fa for disabled mode");
            break;
    }
}


#pragma mark - Reminders

- (nullable NSDate *)lastCompletedReminderDateWithTransaction:(SDSAnyReadTransaction *)transaction
{
    return [OWS2FAManager.keyValueStore getDate:kOWS2FAManager_LastSuccessfulReminderDateKey transaction:transaction];
}

- (void)setLastCompletedReminderDate:(nullable NSDate *)date transaction:(SDSAnyWriteTransaction *)transaction
{
    OWSLogDebug(@"Setting setLastCompletedReminderDate:%@", date);
    [OWS2FAManager.keyValueStore setDate:date key:kOWS2FAManager_LastSuccessfulReminderDateKey transaction:transaction];
}

- (BOOL)isDueForV2ReminderWithTransaction:(SDSAnyReadTransaction *)transaction
{
    if (!self.tsAccountManager.isRegisteredPrimaryDevice) {
        return NO;
    }

    if (!self.hasBackedUpMasterKey) {
        return NO;
    }

    if ([self pinCodeWithTransaction:transaction].length == 0) {
        OWSLogInfo(@"Missing 2FA pin, prompting for reminder so we can backfill it.");
        return YES;
    }

    if (![self areRemindersEnabledTransaction:transaction]) {
        return NO;
    }

    NSDate *nextReminderDate = [self nextReminderDateWithTransaction:transaction];

    return nextReminderDate.timeIntervalSinceNow < 0;
}

- (BOOL)needsLegacyPinMigration
{
    __block BOOL hasMigratedTruncatedPin = NO;
    [self.databaseStorage readWithBlock:^(SDSAnyReadTransaction *transaction) {
        hasMigratedTruncatedPin = [OWS2FAManager.keyValueStore getBool:kOWS2FAManager_HasMigratedTruncatedPinKey
                                                          defaultValue:NO
                                                           transaction:transaction];
    }];

    if (hasMigratedTruncatedPin) {
        return NO;
    }

    // Older versions of the app truncated newly created pins to 16 characters. We no longer do that.
    // If we detect that the user's pin is the truncated length and it was created before we stopped
    // truncating pins, we'll need to ensure we migrate to the user's entire pin next time we prompt
    // them for it.
    if (self.mode == OWS2FAMode_V1 && self.pinCode.length >= kLegacyTruncated2FAv1PinLength) {
        return YES;
    }

    // We don't need to migrate this pin, either because it's v2 or short enough that
    // we never truncated it. Mark it as complete so we don't need to check again.

    DatabaseStorageWrite(self.databaseStorage, ^(SDSAnyWriteTransaction *transaction) {
        [self markLegacyPinAsMigratedWithTransaction:transaction];
    });

    return NO;
}

- (void)markLegacyPinAsMigratedWithTransaction:(SDSAnyWriteTransaction *)transaction
{
    [OWS2FAManager.keyValueStore setBool:YES key:kOWS2FAManager_HasMigratedTruncatedPinKey transaction:transaction];
}

- (void)verifyPin:(NSString *)pin result:(void (^_Nonnull)(BOOL))result
{
    NSString *pinToMatch = self.pinCode;

    switch (self.mode) {
    case OWS2FAMode_V2:
        if (pinToMatch.length > 0) {
            result([pinToMatch isEqualToString:[KeyBackupServiceObjcBridge normalizePin:pin]]);
        } else {
            [self verifyKBSPin:pin
                 resultHandler:^(BOOL isValid) {
                     result(isValid);

                     if (isValid) {
                         DatabaseStorageWrite(self.databaseStorage,
                             ^(SDSAnyWriteTransaction *transaction) { [self setPinCode:pin transaction:transaction]; });
                     }
                 }];
        }
        break;
    case OWS2FAMode_V1:
        // Convert the pin to arabic numerals, we never want to
        // operate with pins in other numbering systems.
        result([pinToMatch.ensureArabicNumerals isEqualToString:pin.ensureArabicNumerals]);
        break;
    case OWS2FAMode_Disabled:
        OWSFailDebug(@"unexpectedly attempting to verify pin when 2fa is disabled");
        result(NO);
        break;
    }
}

- (NSDate *)nextReminderDate
{
    __block NSDate *_Nullable value;
    [self.databaseStorage readWithBlock:^(SDSAnyReadTransaction *transaction) {
        value = [self nextReminderDateWithTransaction:transaction];
    }];
    return value;
}

- (NSDate *)nextReminderDateWithTransaction:(SDSAnyReadTransaction *)transaction
{
    NSDate *lastCompletedReminderDate =
        [self lastCompletedReminderDateWithTransaction:transaction] ?: [NSDate distantPast];
    NSTimeInterval repetitionInterval = [self repetitionIntervalWithTransaction:transaction];

    return [lastCompletedReminderDate dateByAddingTimeInterval:repetitionInterval];
}

- (NSArray<NSNumber *> *)allRepetitionIntervals
{
    // Keep sorted monotonically increasing.
    return @[
        @(12 * kHourSecs),
        @(1 * kDaySecs),
        @(3 * kDaySecs),
        @(7 * kDaySecs),
        @(14 * kDaySecs),
    ];
}

- (double)defaultRepetitionInterval
{
    return self.allRepetitionIntervals.firstObject.doubleValue;
}

- (NSTimeInterval)repetitionInterval
{
    __block NSTimeInterval value;
    [self.databaseStorage readWithBlock:^(SDSAnyReadTransaction *transaction) {
        value = [self repetitionIntervalWithTransaction:transaction];
    }];
    return value;
}

- (NSTimeInterval)repetitionIntervalWithTransaction:(SDSAnyReadTransaction *)transaction
{
    return [OWS2FAManager.keyValueStore getDouble:kOWS2FAManager_RepetitionInterval
                                     defaultValue:self.defaultRepetitionInterval
                                      transaction:transaction];
}

- (void)reminderCompletedWithIncorrectAttempts:(BOOL)incorrectAttempts
{
    DatabaseStorageWrite(self.databaseStorage, ^(SDSAnyWriteTransaction *transaction) {
        [self setLastCompletedReminderDate:[NSDate new] transaction:transaction];

        NSTimeInterval oldInterval = [self repetitionIntervalWithTransaction:transaction];
        NSTimeInterval newInterval = [self adjustRepetitionInterval:oldInterval
                                              withIncorrectAttempts:incorrectAttempts];

        OWSLogInfo(@"Updating repetition interval: %f -> %f. Had incorrect attempts: %d",
            oldInterval,
            newInterval,
            incorrectAttempts);

        [OWS2FAManager.keyValueStore setDouble:newInterval
                                           key:kOWS2FAManager_RepetitionInterval
                                   transaction:transaction];
    });
}

- (NSTimeInterval)adjustRepetitionInterval:(NSTimeInterval)oldInterval withIncorrectAttempts:(BOOL)incorrectAttempts
{
    NSArray<NSNumber *> *allIntervals = self.allRepetitionIntervals;

    NSUInteger oldIndex =
        [allIntervals indexOfObjectPassingTest:^BOOL(NSNumber *_Nonnull obj, NSUInteger idx, BOOL *_Nonnull stop) {
            return oldInterval <= (NSTimeInterval)obj.doubleValue;
        }];

    NSUInteger newIndex;
    if (!incorrectAttempts) {
        newIndex = oldIndex + 1;
    } else {
        // prevent overflow
        newIndex = oldIndex <= 0 ? 0 : oldIndex - 1;
    }

    // clamp to be valid
    newIndex = MAX(0, MIN(allIntervals.count - 1, newIndex));

    NSTimeInterval newInterval = allIntervals[newIndex].doubleValue;
    return newInterval;
}

- (void)setDefaultRepetitionIntervalWithTransaction:(SDSAnyWriteTransaction *)transaction
{
    [OWS2FAManager.keyValueStore setDouble:self.defaultRepetitionInterval
                                       key:kOWS2FAManager_RepetitionInterval
                               transaction:transaction];
}

- (BOOL)areRemindersEnabled
{
    __block BOOL areRemindersEnabled;
    [self.databaseStorage readWithBlock:^(SDSAnyReadTransaction *transaction) {
        areRemindersEnabled = [self areRemindersEnabledTransaction:transaction];
    }];
    return areRemindersEnabled;
}

- (BOOL)areRemindersEnabledTransaction:(SDSAnyReadTransaction *)transaction
{
    return [OWS2FAManager.keyValueStore getBool:kOWS2FAManager_AreRemindersEnabled
                                   defaultValue:YES
                                    transaction:transaction];
}

- (void)setAreRemindersEnabled:(BOOL)areRemindersEnabled transaction:(SDSAnyWriteTransaction *)transaction
{
    return [OWS2FAManager.keyValueStore setBool:areRemindersEnabled
                                            key:kOWS2FAManager_AreRemindersEnabled
                                    transaction:transaction];
}

@end

NS_ASSUME_NONNULL_END

File: SignalServiceKit/src/Util/OWS2FAManager.swift
//
// Copyright 2020 Signal Messenger, LLC
// SPDX-License-Identifier: AGPL-3.0-only
//

import Foundation

extension OWS2FAManager {

    @objc
    public static var isRegistrationLockV2EnabledKey = "isRegistrationLockV2Enabled"

    @objc
    public var isRegistrationLockEnabled: Bool {
        switch mode {
        case .V2:
            return isRegistrationLockV2Enabled
        case .V1:
            return true // In v1 reg lock and 2fa are the same thing.
        case .disabled:
            return false
        }
    }

    @objc
    public var isRegistrationLockV2Enabled: Bool {
        return databaseStorage.read { transaction in
            OWS2FAManager.keyValueStore().getBool(
                OWS2FAManager.isRegistrationLockV2EnabledKey,
                defaultValue: false,
                transaction: transaction
            )
        }
    }

    public func isRegistrationLockV2Enabled(transaction: SDSAnyReadTransaction) -> Bool {
        return OWS2FAManager.keyValueStore().getBool(
            OWS2FAManager.isRegistrationLockV2EnabledKey,
            defaultValue: false,
            transaction: transaction
        )
    }

    public func requestEnable2FA(withPin pin: String, mode: OWS2FAMode, rotateMasterKey: Bool = false) -> Promise<Void> {
        return Promise { future in
            requestEnable2FA(withPin: pin, mode: mode, rotateMasterKey: rotateMasterKey, success: {
                future.resolve()
            }) { error in
                future.reject(error)
            }
        }
    }

    @objc
    @available(swift, obsoleted: 1.0)
    public func enableRegistrationLockV2() -> AnyPromise {
        return AnyPromise(enableRegistrationLockV2())
    }

    public func enableRegistrationLockV2() -> Promise<Void> {
        return DispatchQueue.global().async(.promise) { () -> String in
            let token = Self.databaseStorage.read { tx in
                return DependenciesBridge.shared.svr.data(
                    for: .registrationLock,
                    transaction: tx.asV2Read
                )?.canonicalStringRepresentation
            }
            guard let token else {
                throw OWSAssertionError("Cannot enable registration lock without an existing PIN")
            }
            return token
        }.then { token -> Promise<HTTPResponse> in
            let request = OWSRequestFactory.enableRegistrationLockV2Request(token: token)
            return self.networkManager.makePromise(request: request)
        }.done { _ in
            self.databaseStorage.write { transaction in
                OWS2FAManager.keyValueStore().setBool(
                    true,
                    key: OWS2FAManager.isRegistrationLockV2EnabledKey,
                    transaction: transaction
                )
            }
            firstly {
                return Promise.wrapAsync {
                    try await DependenciesBridge.shared.accountAttributesUpdater.updateAccountAttributes(authedAccount: .implicit())
                }
            }.catch { error in
                Logger.error("Error: \(error)")
            }
        }
    }

    public func markRegistrationLockV2Enabled(transaction: SDSAnyWriteTransaction) {
        guard !DependenciesBridge.shared.tsAccountManager.registrationState(tx: transaction.asV2Read).isRegistered else {
            return owsFailDebug("Unexpectedly attempted to mark reglock as enabled after registration")
        }

        OWS2FAManager.keyValueStore().setBool(
            true,
            key: OWS2FAManager.isRegistrationLockV2EnabledKey,
            transaction: transaction
        )
    }

    @objc
    @available(swift, obsoleted: 1.0)
    public func disableRegistrationLockV2() -> AnyPromise {
        return AnyPromise(disableRegistrationLockV2())
    }

    public func disableRegistrationLockV2() -> Promise<Void> {
        return firstly { () -> Promise<HTTPResponse> in
            let request = OWSRequestFactory.disableRegistrationLockV2Request()
            return self.networkManager.makePromise(request: request)
        }.done { _ in
            self.databaseStorage.write { transaction in
                OWS2FAManager.keyValueStore().removeValue(
                    forKey: OWS2FAManager.isRegistrationLockV2EnabledKey,
                    transaction: transaction
                )
            }
            firstly {
                return Promise.wrapAsync {
                    try await DependenciesBridge.shared.accountAttributesUpdater.updateAccountAttributes(authedAccount: .implicit())
                }
            }.catch { error in
                Logger.error("Error: \(error)")
            }
        }
    }

    public func markRegistrationLockV2Disabled(transaction: SDSAnyWriteTransaction) {
        guard !DependenciesBridge.shared.tsAccountManager.registrationState(tx: transaction.asV2Read).isRegistered else {
            return owsFailDebug("Unexpectedly attempted to mark reglock as disabled after registration")
        }

        OWS2FAManager.keyValueStore().removeValue(
            forKey: OWS2FAManager.isRegistrationLockV2EnabledKey,
            transaction: transaction
        )
    }

    @objc
    @available(swift, obsoleted: 1.0)
    public func migrateToRegistrationLockV2() -> AnyPromise {
        return AnyPromise(migrateToRegistrationLockV2())
    }

    public func migrateToRegistrationLockV2() -> Promise<Void> {
        guard let pinCode = pinCode else {
            return Promise(error: OWSAssertionError("tried to migrate to registration lock V2 without legacy PIN"))
        }

        return firstly {
            return requestEnable2FA(withPin: pinCode, mode: .V2)
        }.then {
            return self.enableRegistrationLockV2()
        }
    }

    @objc
    public func enable2FAV1(pin: String,
                            success: (() -> Void)?,
                            failure: ((Error) -> Void)?) {
        // Convert the pin to arabic numerals, we never want to
        // operate with pins in other numbering systems.
        let request = OWSRequestFactory.enable2FARequest(withPin: pin.ensureArabicNumerals)
        firstly {
            Self.networkManager.makePromise(request: request)
        }.done(on: DispatchQueue.main) { _ in
            Self.databaseStorage.write { transaction in
                self.markEnabled(pin: pin, transaction: transaction)
            }
            success?()
        }.catch(on: DispatchQueue.main) { error in
            owsFailDebugUnlessNetworkFailure(error)
            failure?(error)
        }
    }

    @objc
    public func disable2FAV1(success: OWS2FASuccess?,
                             failure: OWS2FAFailure?) {
        let request = OWSRequestFactory.disable2FARequest()
        firstly {
            Self.networkManager.makePromise(request: request)
        }.done(on: DispatchQueue.main) { _ in
            Self.databaseStorage.write { transaction in
                self.markDisabled(transaction: transaction)
            }
            success?()
        }.catch(on: DispatchQueue.main) { error in
            owsFailDebugUnlessNetworkFailure(error)
            failure?(error)
        }
    }

    public static func isWeakPin(_ pin: String) -> Bool {
        let normalizedPin = SVRUtil.normalizePin(pin)

        guard pin.count >= kMin2FAv2PinLength else { return true }

        // We only check numeric pins for weakness
        guard normalizedPin.digitsOnly() == normalizedPin else { return false }

        var allTheSame = true
        var forwardSequential = true
        var reverseSequential = true

        var previousWholeNumberValue: Int?
        for character in normalizedPin {
            guard let current = character.wholeNumberValue else {
                owsFailDebug("numeric pin unexpectedly contatined non-numeric characters")
                break
            }

            defer { previousWholeNumberValue = current }
            guard let previous = previousWholeNumberValue else { continue }

            if previous != current { allTheSame = false }
            if previous + 1 != current { forwardSequential = false }
            if previous - 1 != current { reverseSequential = false }

            if !allTheSame && !forwardSequential && !reverseSequential { break }
        }

        return allTheSame || forwardSequential || reverseSequential
    }

    @objc
    public func clearLocalPinCode(transaction: SDSAnyWriteTransaction) {
        Self.keyValueStore().removeValue(forKey: kOWS2FAManager_PinCode, transaction: transaction)
    }

    // MARK: - KeyBackupService Wrappers/Helpers

    @objc
    public func hasBackedUpMasterKey(transaction: SDSAnyReadTransaction) -> Bool {
        return DependenciesBridge.shared.svr.hasBackedUpMasterKey(transaction: transaction.asV2Read)
    }

    @objc(generateAndBackupKeysWithPin:rotateMasterKey:)
    @available(swift, obsoleted: 1.0)
    public func generateAndBackupKeys(with pin: String, rotateMasterKey: Bool) -> AnyPromise {
        let promise = DependenciesBridge.shared.svr.generateAndBackupKeys(pin: pin, authMethod: .implicit, rotateMasterKey: rotateMasterKey)
        return AnyPromise(promise)
    }

    @objc
    public func verifyKBSPin(_ pin: String, resultHandler: @escaping (Bool) -> Void) {
        DependenciesBridge.shared.svr.verifyPin(pin, resultHandler: resultHandler)
    }

    @objc(deleteKeys)
    public func deleteKBSKeys() -> AnyPromise {
        return AnyPromise(DependenciesBridge.shared.svr.deleteKeys())
    }
}
